### ISSUE TITLE
Duplicate Users Created at Checkout

### ISSUE THREAD

**Issue by masshoff** [ingenesis/shopp#3628]
Updated to 1.3.12 - now all customers are having new/duplicate accounts/customer ID's created even when they already have an existing account with our Shop.

**Comment by sierrab123** [comment]
I have had a similar report on another ticket

**Comment by scottbuckingham** [comment]
Same happening for me - customer is running a site with hundreds of transactions per day.  The process seems to be:

1. existing user adds item(s) to cart
2. user checks out
3. new user is created in the system with all personal, billing, and shipping information - order is bound to that user
4. existing user no longer has billing or shipping information filled in
5. user is automatically logged in as newly created user

We've updated some orders to point to old users, and have deleted the newly created users (and have copied addresses back), but new users continue to be created.

**Comment by sierrab123** [comment]
https://github.com/ingenesis/shopp/issues/3603

**Comment by jonathandavis** [comment]
I'd like to confirm some additional details on this issue @masshoff and @scottbuckingham. What setting is enabled under Shopp → Setup → Checkout → *Customer Accounts*?

**Comment by masshoff** [comment]
"Enable Account Logins integrated with WordPress Users" is enabled.

**Comment by scottbuckingham** [comment]
We're just using "Enable Account Logins" - actually tried the WordPress option for a bit but that disables login for any existing accounts in the system...

**Comment by jonathandavis** [comment]
It does not appear to be creating duplicate WordPress user logins, just duplicate ShoppCustomer entries.

**Comment by masshoff** [comment]
Jonathan - that is correct. Duplicate Shopp Customers, not duplicate WP users.

**Comment by jonathandavis** [comment]
The machinery that does this is pretty straightforward.

This line of code gets the Customer data from the active session:

```php
$Customer = ShoppOrder()->Customer;
```

The rest of the code is here: https://github.com/ingenesis/shopp/blob/1.3.x/core/flow/Registration.php#L181-L196

The key pieces are checking if the customer already exists. This happens by using the `exists()` method which does a database check to see if `$Customer->id` actually exists as a record in the database. If it does exists, it will return true, if not, it returns false and forces the creation of a new record. If `$Customer->id` is `false` when the customer object is saved to the database, it will generate a new id when creating a new database record. This is the crux of it:

```php
if ( ! $Customer->exists(true) ) {
	$Customer->id = false;
	...
}
```

And the final piece:

```php
$Customer->save();
```

Diagnosing this means that if the `$Customer->exists(true)` returns false, then it's going to create a new Customer record. What we don't know at this time is if the `ShoppCustomer` object available at this time is somehow getting the `id` property set to false, or some other erroneous data value. Any mismatch there will trigger new record creation. This could be caused by session problems, but if the session is already holding all the other `ShoppCustomer` details, why would the `id` property not be correctly kept as well?

In my search, the code above is the only place in all of Shopp that the Shopp Customer `id` property is forced to false, and that is only after it is determined to be a non-existing record.

That leaves a problem in the login/customer record loading mechanism. That is the next area I will explore. I'll try to run tests and see if I can reproduce, otherwise I will need access to an install environment where this is happening in the wild.

It would also be helpful to know what payment system is being used when the Customer duplication occurs (@scottbuckingham and @masshoff). Thanks!

**Comment by masshoff** [comment]
We are using PayPal Pro

**Comment by scottbuckingham** [comment]
We're using PayPal payflow pro and PayPal express.  Any updates on this?

**Comment by scottbuckingham** [comment]
Hey @jonathandavis, any progress on this bug?

**Comment by scottbuckingham** [comment]
@jonathandavis - we need an update on this urgently, please.  Here's our client's situation:

> Can you please advise if there is any progress on this issue.  I am losing business, customers & time.....there are shipping errors happening based on all the different names coming through.
> 
> Some customers can't checkout due to address issues, they cant go back and track their orders. Our online business is built on trust and reliability, we are losing both....

Any update at all... please.

**Comment by sierrab123** [comment]
@scottbuckingham There is a beta fix at Git Hub, if you are interested in trying it (bearing in mind it is not an official release and comes with all those warnings) I have been running on a test site with no issues, you can download it from here https://github.com/ingenesis/shopp/tree/fix/customer-registration

**Comment by scottbuckingham** [comment]
@sierrab123, thanks for providing an update.  What's the release schedule for that fix?  I don't think the client will go for patching a problem with a beta

**Comment by sierrab123** [comment]
I will try and get it out asap, I need to speak to Jon....

**Comment by scottbuckingham** [comment]
Perfect, thanks - we'll wait for the final word!

**Comment by scottbuckingham** [comment]
@sierrab123 @jonathandavis it's been nearly two months since this critical bug has been flagged - what's the status of the public release of the fix you mentioned two weeks ago?

**Comment by masshoff** [comment]
Agreed - an update on the fix would be appreciated.

**Comment by scottbuckingham** [comment]
@sierrab123 @jonathandavis Seriously guys. Release this fix. We can't rely on Shopp, and that means that our customers can't rely on us.

> "It has now been 2 months that we have had to deal with this debilitating issue. I have been more than patient. I understand that this might not be all your fault but you are my only contact to resolve this issue for me. I have had many customers that have stopped buying from us due to this issue and that is unacceptable. I am not prepared to wait anymore. If this issue is not resolved in the next week I will have no other alternative than to look elsewhere for our future needs. I have partners and customers I have to answer to."

**Comment by Just-m** [comment]
You can make it work again, by making the following changes.

**file: /core/library/DB.php replace**

```php
	public function copydata ( $data, $prefix = '', $ignores = false ) {
		if ( ! is_array($ignores) || ! $ignores ) $ignores = array('_datatypes', '_table', '_key', '_lists', '_map', 'id', 'created', 'modified');
```
with
```php
	public function copydata ( $data, $prefix = '', array $ignores = array('_datatypes', '_table', '_key', '_lists', '_map', 'id', 'created', 'modified') ) {
		if ( ! is_array($ignores) ) $ignores = array();
```

**file: /core/flow/Purchased.php replace**
```php
	public function copydata ( $Item, $prefix = '', $ignores = false ) {
		parent::copydata($Item, $ignores);
```
with
```php
public function copydata ( $Item, $prefix = '', array $ignores = array() ) {
		parent::copydata ($Item);
```

Backup the original files for quick revert, in case things go south.

**Comment by Just-m** [comment]
Or just apply this commit https://github.com/ingenesis/shopp/pull/3631

**Comment by Just-m** [comment]
Commit #3631 has been merged.
